code/game/mecha/mech_fabricator.dm

```diff
@@ -172,34 +239,26 @@
 	atom_say("[I] is complete.")
 	being_built = null
 
-	updateUsrDialog()
+	SSnanoui.update_uis(src)
 	return TRUE
 
-/obj/machinery/mecha_part_fabricator/proc/update_queue_on_page()
-	send_byjax(usr,"mecha_fabricator.browser","queue",list_queue())
-	return
-
-/obj/machinery/mecha_part_fabricator/proc/add_part_set_to_queue(set_name)
-	if(set_name in part_sets)
-		for(var/v in files.known_designs)
-			var/datum/design/D = files.known_designs[v]
-			if(D.build_type & fabricator_type)
-				if(set_name in D.category)
-					add_to_queue(D)
-
-/obj/machinery/mecha_part_fabricator/proc/add_to_queue(D)
+// Add a single part to the queue
+/obj/machinery/mecha_part_fabricator/proc/add_to_queue(datum/design/D)
 	if(!istype(queue))
 		queue = list()
-	if(D)
+	if(D && istype(D))
 		queue[++queue.len] = D
-	return queue.len
+		return TRUE
+	return FALSE
 
+// Removes an item from the queue at the specified index
 /obj/machinery/mecha_part_fabricator/proc/remove_from_queue(index)
-	if(!isnum(index) || !istype(queue) || (index<1 || index>queue.len))
+	if(!isnum(index) || !istype(queue) || (index < 1 || index > queue.len))
 		return FALSE
-	queue.Cut(index,++index)
+	queue.Remove(queue[index])

```

Comment:
`.Remove` is 1) more expensive than `.Cut`, 2) does not always do what you expect.

Scenario: the queue has `["left leg", "honkbot", "right leg", "left arm", "right arm"]`.
You then realize that honkbot is probably not as important as the rest, and try to move it to the end of the queue. To do that you add a new honkbot, giving you `["left leg", "honkbot", "right leg", "left arm", "right arm", "honkbot"]`. Then you click "remove" on the earlier one and get up with `["left leg", "honkbot", "right leg", "left arm", "right arm"]`, exactly what you started with.

Bonus points: this also breaks for "normal" queue processing of `["left leg", "honkbot", "right leg", "left arm", "right arm", "honkbot"]` - both honkbots would get finished before right leg is started.

Reply:
>.Remove is 1) more expensive than .Cut, 2) does not always do what you expect.

Perhaps. How do you know this/where did you see this information?

Also I could not get Cut to simply cut a single value. I don't remember exactly why I couldn't get it working, as I changed this early in life of this project. So I switched to remove instead.

But I see the problem you're talking about. I tried adding a duplicate item at the end of the queue and removing the old item further up but it instead removes the one on the end. I have NO idea why this is happening because I'm passing the specific index of the item from the template.

Reply:
> How do you know this/where did you see this information?

The "more expensive" part? Generic algorithmic knowledge.
To remove an item by value from the list you need to inspect each element if it equals that value, and if it does - remove that index.
To remove an item by index - you just remove that index, skipping the "check for equality" part.

Reply:
Now that I take a closer look, I know what you mean, I did `Remove(queue[index])` which is a value instead of just doing `Remove(index)` which, I don't think Remove can even do. So I see why Cut is preferred, you can feed it indexes.